src/stakeinput.cpp

```diff
@@ -18,81 +19,69 @@ CZPivStake::CZPivStake(const libzerocoin::CoinSpend& spend)
+ * Find the first occurrence of a certain accumulator checksum.
+ * Return block index pointer or nullptr if not found
+ */
+
 CBlockIndex* CZPivStake::GetIndexFrom()
 {
-    if (pindexFrom)
-        return pindexFrom;
-
+    // First look in the legacy database
     int nHeightChecksum = 0;
-
-    if (fMint)
-        nHeightChecksum = GetChecksumHeightFromMint();
-    else
-        nHeightChecksum = GetChecksumHeightFromSpend();
-
-    if (nHeightChecksum < Params().Zerocoin_StartHeight() || nHeightChecksum > chainActive.Height()) {
-        pindexFrom = nullptr;
-    } else {
-        //note that this will be a nullptr if the height DNE
-        pindexFrom = chainActive[nHeightChecksum];
+    if (zerocoinDB->ReadAccChecksum(nChecksum, denom, nHeightChecksum)) {
+        return chainActive[nHeightChecksum];
     }
 
-    return pindexFrom;
+    // Not found. Scan the chain.
+    CBlockIndex* pindex = chainActive[Params().Zerocoin_StartHeight()];
+    if (!pindex) return nullptr;
+    const int last_block = Params().Zerocoin_Block_Last_Checkpoint();
+    while (pindex && pindex->nHeight <= last_block) {
+        if (ParseAccChecksum(pindex->nAccumulatorCheckpoint, denom) == nChecksum) {
+            // Found. Save to database and return
+            zerocoinDB->WriteAccChecksum(nChecksum, denom, pindex->nHeight);
+            return pindex;
+        }
+        //Skip forward in groups of 10 blocks since checkpoints only change every 10 blocks
+        if (pindex->nHeight % 10 == 0) {
+            pindex = chainActive[pindex->nHeight + 10];
+            continue;
+        }
+        pindex = chainActive.Next(pindex);
+    }
+    return nullptr;
 }
 
 CAmount CZPivStake::GetValue()
 {
     return denom * COIN;
 }
 
-//Use the first accumulator checkpoint that occurs 60 minutes after the block being staked from
-// In case of regtest, next accumulator of 60 blocks after the block being staked from
+// Use the first accumulator checkpoint that occurs 60 minutes after the block being staked from
 bool CZPivStake::GetModifier(uint64_t& nStakeModifier)
 {
     CBlockIndex* pindex = GetIndexFrom();
     if (!pindex)
         return error("%s: failed to get index from", __func__);
 
     if(Params().IsRegTestNet()) {
-        // Stake modifier is fixed for now, move it to 60 blocks after this pindex in the future..
-        nStakeModifier = pindexFrom->nStakeModifier;
+        nStakeModifier = 0;
         return true;
     }
 
     int64_t nTimeBlockFrom = pindex->GetBlockTime();
-    // zPIV staking is disabled long before block v7 (and checkpoint is not included in blocks since v7)
-    // just return false for now. !TODO: refactor/remove this method
-    while (pindex && pindex->nHeight + 1 <= std::min(chainActive.Height(), Params().Zerocoin_Block_Last_Checkpoint()-1)) {
+    const int nHeightStop = std::min(chainActive.Height(), Params().Zerocoin_Block_Last_Checkpoint()-1);
+    while (pindex && pindex->nHeight + 1 <= nHeightStop) {
```

Comment:
Commenting for the future, no need to do it now.

Start directly pindex + 40 (said 40 to leave some spacing for old block timing issues) or well, check which is the minimum movement and start there.

Reply:
Mhm. Yes, first you should know what is the minimum number of blocks occurred in a 1-hour interval (since zPoS start) and use it here (kind of like a magic number). 
Maybe not worth for just a few checks imo.

Anyway this function is going to be removed from here (and put it back in kernel) in #1291